doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+     * get_<object-type>s(&self) -> Vec<ObjectType>;
+     */
+    /// Get all clusters.
+    fn get_clusters(&mut self) -> Vec<Cluster>;
+
+    /*
+     * Currently, there isn't much need for methods of the form
+     * get_<object-type>_by_id(&self, id: I) -> ObjectType;
+     * As we separate the catalog state out from the rest of the
+     * Coordinator, we will most likely need to add these methods
+     * which should be fairly straight forward.
+     */
+
+    // ...
+
+    // Write
+
+    /// Creates a new durable catalog state transaction. 
+    fn transaction(&mut self) -> Transaction;
+
+    /// Commits a durable catalog state transaction.
+    fn commit_transaction(&mut self, tx: Transaction);
+
+    /*
+     * The majority of writes will go through a Transaction.
+     * However as an optimization it may be useful to have more
+     * targeted methods. We can add those on a case-by-case basis.  
+     */
+
+    /// Persist mapping from system objects to global IDs and fingerprints.
+    fn set_system_object_mapping(
+        &mut self,
+        mappings: Vec<SystemObjectMapping>,
+    );
+
+    // ...
+
+    // Misc
+
+    /// Confirms that this catalog is connected as the current leader.
+    ///
+    /// NB: We may remove this in later iterations of Pv2.
+    fn confirm_leadership(&mut self) -> Result<(), Error>;
+
+    /// Dumps the entire catalog contents in human readable JSON.
+    fn dump(&self) -> String;
+}
+```
+
+We will also add the following methods to initialize the durable catalog state structs. They do
+not open the catalog, but they may create connections over the network.
+
+```Rust
+fn stash_catalog_state(
+    now: NowFn,
+    stash_factory: StashFactory,
+    stash_url: String,
+    tls: MakeTlsConnector,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+
+fn persist_catalog_state(
+    now: NowFn,
+    shard_id: ShardId,
+    persist_client: PersistClient,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+
+fn shadow_catalog_state(
+    primary_state: Box<dyn DurableCatalogState>,
+    secondary_state: Box<dyn DurableCatalogState>,
+) -> Box<dyn DurableCatalogState> {
+    // ...
+}
+```
+
+These methods are needed to solve a bootstrapping problem. Some catalog state, such as deploy
+generation, is needed before attempting to open the catalog. So these methods will return unopened
+catalog states capable of returning only the state we need for bootstrapping.
+
+All current usages of `mz_adapter::catalog::storage::Connection` will be updated to
+use `Box<dyn DurableCatalogState>`. The `environmentd` binary will initialize
+the `Box<dyn DurableCatalogState>` using the methods above depending on command line arguments.
+
+### Stash implementation
+
+A stash based implementation will be fairly trivial to implement. This interface is already
+implemented using the stash, we will just need to change the name of some methods.
+
+### Persist Implementation
```

Comment:
Once the persist impl is complete, I'm going to create benchmarks comparing the persist impl to the stash impl.

@danhhz do you have thoughts on how a persist write using `compare_and_append_batch` compares in performance to a CRDB read-write txn?

Reply:
> I'm not sure that inline writes would be the limiting factor there as long as we're only doing one persist write (is there only one?). what other sorts of user operations involve catalog writes?

- DDL involves two catalog writes, one for allocating the ID and one for adding the object. Though we should be able to combine these into a single write.
- Table writes sometimes involve a catalog write (for the timestamp oracle).
- Strict serializable reads will involve an unnecessary catalog write to confirm leadership for the stash impl that I'm trying to avoid in the persist impl.
- Initialization/migration involves a whole bunch of catalog writes. Many of these can probably be combined into a single write. Though since this happens so infrequently I think it's OK if it ends up a bit slower by some ms.